src/mcts/node.cc

```diff
@@ -417,8 +482,12 @@ bool NodeTree::ResetToPosition(const std::string& starting_fen,
   // previously trimmed; we need to reset current_head_ in that case.
   // Also, if the current_head_ is terminal, reset that as well to allow forced
   // analysis of WDL hits, or possibly 3 fold or 50 move "draws", etc.
-  if (!seen_old_head || current_head_->IsTerminal()) TrimTreeAtHead();
+  if (!seen_old_head) TrimTreeAtHead(); 
 
+  // Certainty Propagation: No need to trim the head, just resetting certainty
+  // state except bounds, and recomputing N should suffices even with WDL hits.
+  if (current_head_->GetParent()) current_head_->GetOwnEdge()->ClearCertaintyState();
+  current_head_->RecomputeNfromChildren();
```

Comment:
I'm not seeing how this works - terminals and certains created at ExtendNode time don't get their edges generated - so when they become the head they have no edges - which breaks the concept of choosing the next move from that position.
If RecomputeNFromChildren could set it to 0 if it has no children, that might work - but since it sets n to 1 it won't be candidate for selection for extension.  Generally I think if current_head_ has no child edges, its safer to call TrimTreeAtHead like we used to.

Reply:
The rational is this:
(1) certain edges are not only created in ExtendNode but also when backpropagating up the tree 
(2) as soon as a normal edge is made certain through backpropagation the children of the node it leads to no longer get visits, while the node itself might still get visits.
(3) Now, if such a node becomes the new head, the sum of its children+ 1 is no longer consisten with its N. 

RecomputeNfromChildren ensures this consistency and only changes n if n>1, so for node with no children N=0. 

At least thats whats supposed to happen.

Reply:
Yes I understand what you are trying to do -  and it probably makes sense for certain edges that aren't created in ExtendNode - but those that are created in ExtendNode don't get their own edges added, but do still have GetN() returning something other than 0..  If you don't fix that they can't be used as a valid root node.

Reply:
I see the following cases potentially arising. Keep in mind that RecomputeNfromChildren only recalculates when the conditional ´if (n_ > 1) ´ is met, otherwise it leaves n_ untourched:

(1) A nodes is extended, its child edges get created, the NN results are fetched and its score updated. After this n_ = 1 and the conditional (n_ > 1) is not met.  If this node becomes root
->  behaviour from normal leela is unaltered, so n_ is correct and consistent

(2) A node is extended, its child edges get createt, but the NN results are not yet fetched. This case probably can't even happen, but to illustrate. If this node gets to be root:
-> behaviour from normal leela is unaltered, so n_ is correct and consistent.

(3) A (non certain) node has already a visit, and child edges and now gets a second visit that leads to extending one of its childs. Here this node has n_ = 2, and one of the children has n_1. If this node gets to be root,the conditional kicks and n_ is recomputed to the same value n_= 2.  
-> so, n_ is correct and consistent.
This should hold for any number of prior visits.

We only really need to Recompute if the node is certain, maybe such a condition could be added for optimization.  But apart from that I must be missing something here? Please explain.

Reply:
I added a separate comment to maybe draw your attention to the specific problem more clearly.  You seem to be missing the case I was specifically calling out.